external/fv3fit/fv3fit/emulation/microphysics/layers.py

```diff
@@ -0,0 +1,267 @@
+
+    def call(self, tensors):
+
+        field_input, network_output = tensors
+        tendency = self.tendency(network_output)
+        tensor = self.increment([field_input, tendency])
+
+        if self.use_relu:
+            tensor = self.relu(tensor)
+
+        return tensor
+
+    def get_tendency_output(self, network_output):
+        return self.tendency(network_output)
+
+
+class CombineInputs(tf.keras.layers.Layer):
+    """Input tensor stacking with option to add a dimension for RNNs"""
+
+    def __init__(self, combine_axis: int, *args, expand: bool = False, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.combine_axis = combine_axis
+        self.expand = expand
+
+    def call(self, inputs):
+
+        if self.expand:
+            inputs = [
+                tf.expand_dims(tensor, axis=self.combine_axis) for tensor in inputs
+            ]
+
+        return tf.concat(inputs, axis=self.combine_axis)
+
+
+def _combine_initializer(do_combine, axis, expand, name):
+
+    if do_combine:
+        layer = CombineInputs(axis, expand=expand)
+    else:
+        layer = tf.keras.layers.Lambda(lambda x: x, name=f"comb_passthru_{name}")
+    return layer
+
+
+class RNNBlock(tf.keras.layers.Layer):
+    """
+    RNN connected to an optional MLP for prediction
+    
+    Combines multiple tensors along a new trailing dimension and
+    operates in the reverse on the feature dimension, i.e., from
+    the top of the atmosphere to the bottom for microphysics data.
+    """
+
+    def __init__(
+        self,
+        *args,
+        channels: int = 256,
+        dense_width: int = 256,
+        dense_depth: int = 1,
+        combine_inputs: bool = True,
+        activation: str = "relu",
+        **kwargs,
+    ):
+        """
+        Args:
+            channels: width of RNN layer
+            dense_width: width of MLP layer, only used if dense_depth > 0
+            dense_depth: number of MLPlayers connected to RNN output, set to
+                0 to disable
+            combine_inputs: whether to stack the inputs into a single block
+            activation: activation function to use for RNN and MLP
+        """
+        super().__init__(*args, **kwargs)
+
+        self.combine = _combine_initializer(combine_inputs, -1, True, self.name)
+        self.rnn = tf.keras.layers.SimpleRNN(
+            channels, activation=activation, go_backwards=True
+        )
+        self.dense = MLPBlock(
+            width=dense_width, depth=dense_depth, combine_inputs=False
+        )
+
+    def call(self, inputs):
+        combined = self.combine(inputs)
+        rnn_out = self.rnn(combined)
+
+        output = self.dense(rnn_out)
+
+        return output
+
+
+class MLPBlock(tf.keras.layers.Layer):
+    """MLP layer for basic NN predictions"""
+
+    def __init__(
+        self,
+        *args,
+        width: int = 256,
+        depth: int = 2,
+        combine_inputs: bool = True,
+        activation: str = "relu",
```

Comment:
There's the subtle difference of built tensor trace vs. a layer used to build a tensor trace, but we should be able to facilitate both if we deside whether layers work as a nice base unit for model building.  Since the args here are nearly a subset of the config, I think I'll leave as is for now and handle in the other PR if we decide to go that route.